// text-runner/dist/formatters/formatter.js
//      

                                                     

const { bold, green, magenta } = require('chalk')
const Time = require('time-diff')
const UnprintedUserError = require('../errors/unprinted-user-error')

                
                         
 

// Base class for formatters
class Formatter {
                      
                  
                 
                      
                   // the path of the documentation file that is currently processed
                      // the files encountered so far
                      // whether this formatter is currently processing an action
                    // the line within the documentation file at which the currently processed block starts
                     
                     
                    
            
                        
                       

  constructor () {
    // Note: I have to define these attributes here,
    //       since doing so at the class level
    //       binds them to the class scope for some reason
    this.activityText = ''
    this.errorMessage = ''
    this.filePaths = []
    this.inActivity = false
    this.stepsCount = 0
    this.warningsCount = 0
    this.stdout = { write: this.output }
    this.stderr = { write: this.output }
    this.console = {
      log: (text        ) => {
        this.output(`${text}\n`)
      }
    }

    this.time = new Time()
    this.time.start('formatter')
  }

  // Called on general errors
  error (errorMessage        ) {
    this.errorMessage = errorMessage
    this.inActivity = false
  }

  output (text                 )          {
    throw new Error('Implement in subclass')
  }

  setLines (startLine        , endLine        ) {
    this.startLine = startLine
    this.endLine = endLine
  }

  skip (activityText        ) {
    if (activityText) this.activityText = activityText
    this.inActivity = false
  }

  // Called when we start performing an activity that was defined in a block
  setTitle (activityText        ) {
    this.activityText = activityText
  }

  startActivity (activityTypeName        ) {
    if (this.inActivity) {
      throw new UnprintedUserError('already in a started block')
    }
    this.setTitle(activityTypeName)
    this.errorMessage = ''
    this.warningMessage = ''
    this.stepsCount += 1
    this.inActivity = true
  }

  // called when we start processing a markdown file
  startFile (filePath        ) {
    this.filePath = filePath
    if (!this.filePaths.includes(filePath)) {
      this.filePaths.push(filePath)
    }
  }

  // called when the last started activity finished successful
  // optionally allows to define the final text to be displayed
  success (activityText         ) {
    if (activityText) this.activityText = activityText
    this.inActivity = false
  }

  // called when the whole test suite passed
  suiteSuccess () {
    if (this.stepsCount === 0) {
      this.warning('no activities found')
      return
    }
    var text = green(
      `\nSuccess! ${this.stepsCount} blocks in ${this.filePaths.length} files`
    )
    if (this.warningsCount > 0) {
      text += green(', ')
      text += magenta(`${this.warningsCount} warnings`)
    }
    text += green(`, ${this.time.end('formatter')}`)
    console.log(bold(text))
  }

  // Called on general warnings
  warning (warningMessage        ) {
    this.warningMessage = warningMessage
    this.warningsCount += 1
    this.inActivity = false
  }
}

module.exports = Formatter
