// text-runner/dist/commands/run/link-target-list-builder.js
//      

                                                                 
                                                           

const dashify = require('dashify')

class LinkTargetListBuilder {
  // determines which files contain which link targets (anchors)

  // the LinkTarget list that this builder is supposed to populate
                             

  constructor (value                                 ) {
    this.linkTargets = value.linkTargets
  }

  buildLinkTargets (filePath        , tree             ) {
    this.linkTargets[filePath] = this.linkTargets[filePath] || []
    for (let node of tree) {
      switch (node.type) {
        case 'htmltag':
          if (node.content != null) {
            const matches = node.content.match(/<a name="([^"]*)">/)
            if (matches != null) {
              this.linkTargets[filePath].push({
                type: 'anchor',
                name: matches[1]
              })
            }
          }
          break

        case 'heading':
          this.linkTargets[filePath].push({
            type: 'heading',
            name: dashify(node.content),
            text: node.content,
            level: node.level
          })
          break
      }
    }

    return tree
  }
}

module.exports = LinkTargetListBuilder
